### PR TITLE
Introduce Drop In UI Module

### PR DESCRIPTION
**Pull request by abhishek1508**
### Description

Fixes https://github.com/mapbox/navigation-sdks/issues/1259 Introduce Drop In UI basic module setup.

### Changelog

```
<changelog>:feature: Introduced Drop In UI basic module setup.</changelog>
```

### CODE REVIEW

**Comment by abhishek1508** [comment]
@LukasPaczos Can I request you to make sure that merging this to main will not add a new release artifact to SDK Registry as we are not ready to release this module yet?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewOptions.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewOptions.kt:47]:
```suggestion
        private val lightColors = Colors(
            light = 100,
            warmth = 100,
            contrast = 0,
            saturation = 0,
            primary = 0,
            secondary = 0,
            background = 0,
            onPrimary = 0,
            onSecondary = 0
        )
        private val darkColors = Colors(
            light = 0,
            warmth = 0,
            contrast = 0,
            saturation = 0,
            primary = 0,
            secondary = 0,
            background = 0,
            onPrimary = 0,
            onSecondary = 0
        )
```

File: settings.gradle
Comment (kmadsen) [inline settings.gradle:23]:
```suggestion
        ':libnavui-dropin',
        ':qa-test-app'
```

nit: qa-test-app can stay at the bottom IMO 😄

File: gradle/dependencies.gradle
Comment (kmadsen) [inline gradle/dependencies.gradle:118]: pretty sure you can remove this when you're using lifecycle `2.4.0` instead of `2.3.1`.

https://developer.android.com/jetpack/androidx/releases/lifecycle#version_24_2

File: libnavui-dropin/build.gradle
Comment (RingerJK) [inline libnavui-dropin/build.gradle:15]: need set up dokka

Comment (abhishek1508) [reply libnavui-dropin/build.gradle:15]: Will set up dokka later once we have a better understanding of how to define public API's

Comment (RingerJK) [reply libnavui-dropin/build.gradle:15]: then need to remove `dokka plugin` and `kdoc-setting` at line `5` and `10` 😄

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt
Comment (RingerJK) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt:12]: why do we use AndroidViewModel, but not just ViewModel? It adds a bit of magic (you have access to appContext), but it becomes harder to write unit-test.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:21]: It seems the way to add a view to this container in a consistent way, would be to add a new boolean visibility flag. For example, when we add the current-road-label - does it go here?

I think I prefer independent streams for visibility. You will also have independent streams for providing data to the containers, and those streams will impact the visibility. For example, speed limit my be visible to the container but is invisible when the road does not have a speed limit value.

I can see this is only connecting, NavigationState visibility. But the streams that control visibility can easily depend on NavigationState rather than defining a new action type

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/ContainerVisibilityViewModel.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/ContainerVisibilityViewModel.kt:20]: This duplicates the `_navigationState: MutableStateFlow<NavigationState>` in the `NavigationStateViewModel`

In reality, this is a transformation of the `Flow<NavigationState>`. The ViewModel here appears to be overkill

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:171]: Can you hook up this part? Reason I'm saying this, is because how are you connecting the NavigationState flow to the containerVisibilityViewModel state flow?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:21]: Here's another idea

``` kotlin
    val speedLimitVisibilitySet = setOf(NavigationState.FreeDrive, NavigationState.RoutePreview)
    fun speedLimitVisible(): Flow<Boolean> = navigationState().map {
        speedLimitVisibilitySet.contains(it)
    }
```

Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:21]: I like the suggestion. I am going to create a new issue for it where I am gonna map the view visibility state to navigation state. Let's not block this PR, as what you are suggesting is an improvement and a good one but the PR unblocks us from moving further on action items.

File: examples/src/main/AndroidManifest.xml
Comment (LukasPaczos) [inline examples/src/main/AndroidManifest.xml:100]: Let's add a separate activity. Also, why not add it to the QA module since `exmaples` goes away anyway?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:171]: I am not connecting the `NavigationState` to container visibility just yet, because we still don't know what needs to happen based on each of these states. I would leave it blank for now but would add more code once we have a clear understanding.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/ContainerVisibilityViewModel.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/ContainerVisibilityViewModel.kt:20]: I agree and as explained above will fix this in a separate PR so as to unblock this.

**Review by kmadsen — APPROVED** [review]
Suggesting we delete `ContainerVisibilityViewModel` and only use NavigationState flow in the comments. Approving because adding this PR is adding a module that everyone can contribute too, also we can always iterate!

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:80]: Let's add default values here.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:77]: Let's default to `null`.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:47]:
```suggestion
    val mapView: MapView
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:62]:
```suggestion
    lateinit var mapboxNavigation: MapboxNavigation
      private set
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationView.kt:144]: Let's use field accessors instead of functions.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt
Comment (abhishek1508) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt:12]: In this case I need context to create an instance of `MapboxNavigation` which is unavailable via the `ViewModel`. We don't want to do that in the view or else it would be destroyed. Do you have any other suggestions to instantiate the `MapboxNavigation` inside `NavigationViewModel` but still use `ViewModel` and get access to context?
cc @RingerJK

File: libnavui-dropin/build.gradle
Comment (abhishek1508) [reply libnavui-dropin/build.gradle:13]: Done.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt:16]: Since we plan to control the instance, we should not rely on it being already created. Drop-In UI should reset state of previous instances and spin up it's own. Since it will live in a view model, it will not be recreated on configuration changes which is enough for us.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:7]:
```suggestion
internal sealed class ContainerVisibilityAction {
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:15]:
```suggestion
internal sealed interface ContainerVisibilityResult {
```

Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:15]: Same below, possible all of this should be internal.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/ContainerVisibilityContract.kt:7]: All of this file should be internal, shouldn't it?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/NavigationStateContract.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/contract/NavigationStateContract.kt:16]: Results/Flows should be internal.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/processor/ContainerVisibilityProcessor.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/processor/ContainerVisibilityProcessor.kt:5]: This should also be internal.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/processor/NavigationStateProcessor.kt
Comment (LukasPaczos) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/processor/NavigationStateProcessor.kt:6]: Same as above.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/ContainerVisibilityViewModel.kt
Comment (LukasPaczos) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/viewmodel/ContainerVisibilityViewModel.kt:20]: I agree that additional view model might be an overkill. The visibility value will ultimately be a transformation of state, visibility settings from config, and data availability.

**Review by LukasPaczos — COMMENTED** [review]
As long as the module is not added to `publish.gradle`, and the module does not add `publish.gradle` into scripts list, the artifact will not be published.

However, let's add the module to the Makefile and to the group of UI modules so that the test, lint, and any static analysis is run.

File: examples/src/main/AndroidManifest.xml
Comment (abhishek1508) [reply examples/src/main/AndroidManifest.xml:100]: The decision was made not to add any test activity to QA test app.

**Comment by abhishek1508** [comment]
Thanks for taking a look @LukasPaczos. This is ready for another round of 👀

**Review by LukasPaczos — APPROVED** [review]
There are still things that need to be cleaned up here but let's use this as a skeleton and iterate :ship:

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt
Comment (RingerJK) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt:12]: post-merge comment 😄 
We can make an application provider object, that would be included to `NavigationViewModel` only as an option

Comment (cafesilencio) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/NavigationViewModel.kt:12]: I have a draft PR where I've done some work on `NavigationViewModel`. I've removed the reference for the Application reference as I don't think it's needed.  It's still subject to review once the PR is ready.

**Pull request merged**
